compiler/rustc_mir/src/transform/early_otherwise_branch.rs

```diff
@@ -46,6 +46,10 @@ impl<'tcx> MirPass<'tcx> for EarlyOtherwiseBranch {
         let should_cleanup = !opts_to_apply.is_empty();
 
         for opt_to_apply in opts_to_apply {
+            if !tcx.consider_optimizing(|| format!("EarlyOtherwiseBranch {:?}", &opt_to_apply)) {
```

Comment:
Yea, as you stated, we should either doe the mutations after we know the optimization should apply, or we should not even do the analysis for whether it should be applied.

Reply:
Well, I'm not 100% sure of that. It seems from @wesleywiser's comment that we should check fuel at the latest possible time (right before the optimization is applied), since the primary purpose is for debugging, not performance. If that's true, then in this case I think it's in the right place, but in `const_prop` and `instcombine` it is applied too early